testing/tests/DevExpress.ui.widgets.editors/tagBox.tests.js

```diff
@@ -5071,6 +5071,75 @@ QUnit.module('the \'fieldTemplate\' option', moduleSetup, () => {
     });
 });
 
+
+QUnit.module('the \'valueChangeEvent\' option', moduleSetup, () => {
+    const events = ['keyup', 'blur', 'change', 'input', 'focusout'];
+    const eventCouples = [
+        'keyup blur',
+        'keyup change',
+        'keyup input',
+        'keyup focusout',
+        'blur change',
+        'blur input',
+        'blur focusout',
+        'change input',
+        'change focusout',
+        'input focusout',
+    ];
+
+    const valueChangeEventValues = [ events, eventCouples ];
+
+    valueChangeEventValues.forEach((events) => {
+        events.forEach((eventValue) => {
+            QUnit.test(`custom item has been added when valueChangeEvent='${eventValue}'`, function(assert) {
+                const onCustomItemCreatingSpy = sinon.spy();
+
+                const $tagBox = $('#tagBox').dxTagBox({
+                    items: ['item 1'],
+                    acceptCustomValue: true,
+                    valueChangeEvent: eventValue,
+                    onCustomItemCreating(args) {
+                        onCustomItemCreatingSpy();
+                        args.customItem = args.text;
+                    },
+                });
+
+                const $input = $tagBox.find(`.${TEXTBOX_CLASS}`);
+                const keyboard = keyboardMock($input);
+                const customValue = 't';
+                // const [first, second] = eventValue.split(' ');
+                const eventsArray = eventValue.split(' ');
+
+                switch(eventValue) {
```

Comment:
Let's implement separate tests to simplify debug and future support. We can move initialization to `beforeEach` to reduce copypaste.

Reply:
https://github.com/DevExpress/DevExtreme/pull/22772/commits/d60c77cdb5d8d06e7b00552fa2aaa61a62a1a3f0